Protect image inside MemoryCache::destroyDecodedDataForAllImages()

cdumez opened the pull request:
#### cdbf599aeb5ffbc458b1f6958888d7675fdefb9f
<pre>
Protect image inside MemoryCache::destroyDecodedDataForAllImages()
<a href="https://bugs.webkit.org/show_bug.cgi?id=258946">https://bugs.webkit.org/show_bug.cgi?id=258946</a>
&lt;rdar://111807064&gt;

Reviewed by Simon Fraser.

Protect image inside MemoryCache::destroyDecodedDataForAllImages() before
calling destroyDecodedData() on it, to make sure the Image stays alive for the
duration of the function execution.

Also do more protecting in MemoryCache::forEachResource() for good measure.

This may explain the crashes at &lt;rdar://111807064&gt;.

* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::destroyDecodedDataForAllImages):

Canonical link: <a href="https://commits.webkit.org/265857@main">https://commits.webkit.org/265857@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/761bd826de7b81d7012d37f4b3c2bdcbe32bbae4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12069 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12415 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12717 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13815 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11648 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12088 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14831 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12429 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14343 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12233 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10216 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14230 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10332 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10953 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18081 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11412 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11113 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14290 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11603 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9558 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10817 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2956 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15142 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11454 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/loader/cache/MemoryCache.cpp:
```diff
@@ -221,12 +221,14 @@ void MemoryCache::pruneLiveResources(bool shouldDestroyDecodedDataForAllLiveReso
-                continue;
-            function(*resource);
-        }
+    Vector<WeakPtr<CachedResource>> allResources;
```
Would be nice if we could reserve some initial capacity here.

cdumez committed:
Protect image inside MemoryCache::destroyDecodedDataForAllImages()
https://bugs.webkit.org/show_bug.cgi?id=258946
<rdar://111807064>

Reviewed by Simon Fraser.

Protect image inside MemoryCache::destroyDecodedDataForAllImages() before
calling destroyDecodedData() on it, to make sure the Image stays alive for the
duration of the function execution.

Also do more protecting in MemoryCache::forEachResource() for good measure.

This may explain the crashes at <rdar://111807064>.

* Source/WebCore/loader/cache/MemoryCache.cpp:
(WebCore::MemoryCache::destroyDecodedDataForAllImages):

Canonical link: https://commits.webkit.org/265857@main
